src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -934,6 +956,32 @@ internal void SkipNewlines(bool skipSemis, bool v3)
                         goto again;
                     }
 
+                    // If the first non-newline, non-whitespace, non-comment, non-backtick, non-semi-colon character
+                    // following the newline is not a pipe, remove the tokens that were added while tokenizing and unget
+                    // the characters that were retrieved. This allows us to "lookahead" to see if there is a pipe and
+                    // only apply line continuance automatically when one is found.
+                    if (skipOption == NewlineSkipOption.BeforePipe && c != '|')
+                    {
+                        var newTokenCount = TokenList != null ? TokenList.Count : 0;
+                        if (newTokenCount > tokenCount)
+                        {
+                            if (tokenCount == 0)
+                            {
+                                TokenList.Clear();
+                                TokenList = null;
+                            }
+                            else
+                            {
+                                TokenList.RemoveRange(tokenCount - 1, newTokenCount - tokenCount + 1);
+                            }
+                        }
+
+                        while (_currentIndex > startIndex)
+                        {
+                            UngetChar();
+                        }
```

Comment:
Have you compared performance? I have a feeling this will be a hotspot in comment-heavy code.

Reply:
IIRC, you might also see extra memory usage (compared to before this change) from the rescan when re-parsing a script for it's help because the comment tokens will be created twice.

---
In reply to: [259051896](https://github.com/PowerShell/PowerShell/pull/8938#discussion_r259051896) [](ancestors = 259051896)

Reply:
Thanks @lzybkr. I haven't compared performance yet, but noticed in the comments that you (or someone on the PS team) seems to have done some significant performance testing using files from Poshcode, Technet, etc. with comments. Are those tests available somewhere?

Wrt both performance and memory usage, I was actually working on refactoring part of the logic partly because CodeFactor was complaining about the complexity of the SkipNewlines method (which I wasn't quite happy with either), and partly because I didn't like the tokens getting created unnecessarily just for lookahead logic (that's the only memory impact, I believe). I also didn't like using a while loop to move the character marker, and instead simply set it back to where it was if the lookahead fails. I just need to figure out why this new logic isn't syntax highlighting the second `gcm`command in this script, and add a test to make sure that works properly:

```
gcm Get-Command
gcm Get-Help
```

Reply:
> Are those tests available somewhere?

Jason put some helpers in tools/performance folder.

Reply:
The corpus you've seen reference to includes many internal scripts that can't be shared and honestly those scripts were maybe the most valuable in that corpus.

Today I'd look to collecting scripts from the PowerShell Gallery and GitHub. (Google BigQuery reports >400,000 scripts on GitHub, though it looks like it reports many duplicates from forks despite the claim that they exclude forks.)

Reply:
Another question. For backtick we create `NewToken(TokenKind.LineContinuation)` - should we create the same for automatic line continuation?

Reply:
I'm not sure. Some tools (e.g. a code formatter) might insert the backtick.

Reply:
For consistency it could be `TokenKind.AutoLineContinuation`.

Reply:
Yeah, anyone want to put in the effort to collect a safe subset of scripts from the internet for performance testing? 😉 

I mean, most scripts on github are likely to be build scripts (not very good for perf testing of PowerShell, probably) or actually just functions that are part of a module -- but there are still tens of thousands of [PowerShell gists](https://gist.github.com/search?utf8=%E2%9C%93&q=language%3APowerShell+if&ref=searchresults) too ...

Reply:
@Jaykul I think we only need to test the parser performance in this case - not actually running them.

Reply:
Very good point 😊

Reply:
I'll run a bunch of tests once I get back to this PR. I decided to put the finishing touches on tests for my PR with the `S.M.A.PowerShell.InvokeAsync` methods first to clear my head a little before digging in deeper in tests with this PR.

Reply:
I added WIP until we get new commits.

Reply:
Ok, I've moved forward on this PR, and have a few comments:

1. @iSazonov: I don't think we should have a token for automatic line continuation. A backtick is a physical token. All other tokens are physical tokens in a script. In this case (automatic line continuation), there is no token to capture, so I think having a new token where there is nothing in the script would just confuse things.
2. @lzybkr: I completely refactored the logic I was using and now just lookahead without tokenizing to determine if there is continuation or not. That fixed the tokenization problem I couldn't put my finger on, plus it keeps memory constant, and it's very fast. Given the change in code, I would like you to have another look. Initial performance tests show no noticeable impact in performance (tested parsing the allcommands.ps1 file from the dbatools module, using the v2 parser and the S.M.A.L.Parser ParseFile method).

Reply:
Ok, I've moved forward on this PR, and have a few comments:

1. @iSazonov: I don't think we should have a token for automatic line continuation. A backtick is a physical token. All other tokens are physical tokens in a script. In this case (automatic line continuation), there is no token to capture, so I think having a new token where there is nothing in the script would just confuse things.
2. @lzybkr: I completely refactored the logic I was using and now just lookahead without tokenizing to determine if there is continuation or not. That fixed the tokenization problem I couldn't put my finger on, plus it keeps memory constant, and it's very fast. Given the change in code, I would like you to have another look. Initial performance tests show no noticeable impact in performance (tested parsing the allcommands.ps1 file from the dbatools module, using the v2 parser and the S.M.A.L.Parser ParseFile method). I'm working out more performance tests now.

Reply:
More performance testing results:

I created a 77MB script file with a lot of block comments containing Lorem Ipsum text and line comments, with no blank lines (intended to be an extreme/worst case scenario based on the lookahead logic), and ran it through the parser (new and 2.0) before and after these changes. Before the changes I get parse times around 8.7s for the file, regardless of which parser I use. After the changes I get parse times ranging from 9.4s up to 10s.

I then created a simple function with the example comment-based help that is in the docs at the end of the function, and copied/pasted the function hundreds of times in the same file to create another 77MB script file, and ran it through the parser (new and 2.0) before and after these changes, and timed the parsers using Measure-Object. The times were identical in this case.

One thought I'd like an opinion on that would reduce the likelihood of this causing any performance issue whatsoever for anyone: with backtick and pipe at the end of the line, we have an explicit line continuance, and can continue our command even after multiple blank lines. With "implicit" (for lack of a better term) line continuance like this, currently it works the same way, allowing scripters to apply multiple blank lines if they choose to do so. Should it? Or should we have the parser enforce that no blank lines or lines consisting of nothing but whitespace are allowed with implicit/automatic line continuance? I think it is very unlikely that someone would want a blank line when using this -- comments, sure, but not blank lines. In testing (I already threw together an experimental build that identifies blank lines or lines containing nothing but whitespace as command breaks when it comes to automatic line continuance), I took the same scrpit I used in the first file and generated a new file with blank lines before the block comments. The result was that pre-automatic line continuance, the file parses in around 8.8s, and post-automatic line continuance, the file parses in around 9.4s, which is a nice improvement over the results of the first test performed above.

@lzybkr, @iSazonov: Any thoughts on whether or not blank/empty lines should allow automatic line continuance or not?

Reply:
> With "implicit" (for lack of a better term) line continuance like this, currently it works the same way, allowing scripters to apply multiple blank lines if they choose to do so. Should it? 

It is tricky question :-) We have different behavior with PSReadline and without it. Perhaps @mklement0  would be interested to see it. Personally, I think PSReadline behavior is more correct.

As for performance tests, the use of Measure-Command does not give a reliable result, although it clearly shows the absence of performance degradation. We could get a more accurate result using WPR/WPA, but it only makes sense after moving to .Net Core 3.0. I hope to find time to drive more carefully the new code, but for now I like the result obtained and I think we could merge before moving to .Net Core 3.0.

Reply:
I agree that the PSReadline behaviour is the correct one. I'm not sure my original question was clear though. When applying automatic line continuance, should we allow blank/whitespace lines in a command or not.

e.g. If someone enters this command, should it work?

```PowerShell
# Get the current process
gps -id $pid

# And pull out its name
| foreach Name
```

Or, should that only work when there is no blank/whitespace line in the middle (comment lines are allowed though), like this:

```PowerShell
# Get the current process
gps -id $pid
# And pull out its name
| foreach Name
```

Mulling this over on the weekend I've found myself thinking blank lines should not be allowed, to keep multi-line commands contiguous in scripts where they are used. I'm also hoping to put in an experimental feature allowing parameter wrapping, and for that I wouldn't want blank lines used either.

e.g. Parameter wrapping (experimental since it would be breaking for the unusual case where you have a command that starts with "-", requiring you to invoke such commands with the call operator).

```PowerShell
Get-Command
    -Name Invoke-Command
    -Syntax
    @AdditionalSplattedParametersIfDesired
```

That's also a contrived example, but the point is to offer easier line wrapping for commands without requiring users to use splatting unless they want/need to.

Reply:
I'd say that blank lines should not be allowed in implicit command continuations. It smells to me of the classic `if`/`else` statements without braces problem in other languages:

```csharp
if (!something)
    // DoThisThingToDealWithIt(); 
DoTheNextThing();
// Oops...
```

Anyway, I'm super excited to see this feature make it into PowerShell. It's annoyed me for a long time that a command cannot be continued with a pipe on a new line, letting the code line up nicer and look as nice as F#.

Reply:
My thoughts is the same - extra lines is a way to tricky errors.